File: game_dutch_wachtwoord.py
import random

from template import Template
from game_dutch import Dutch
from variable import Variable
from unit_test import UnitTest


class Wachtwoord(Dutch):
    def __init__(self):
        super().__init__()
        self.description = 'WACHTWOORD - Bekijk of een wachtwoord voldoende sterk is.'
        self.introduction_template = Template(
            'Introductie',
            'Een webshop laat een nieuwe functie ontwikkelen door een extern softwarebedrijf.',
            'De functie wordt gebruikt om wachtwoorden te controleren voor de webshop.',
            'Gebruikers moeten erop kunnen rekenen dat de getoonde vereisten correct zijn en er mogen dus geen fouten in zitten.',
            'Daarom hebben ze jou ingehuurd om unit testen te schrijven voor die functie.',
            'In het contract vind je wat je moet doen en hoeveel je daarvoor betaald krijgt.',
            'In de specificatie staat wat de functie precies moet doen.',
        )
        self.specification_template = Template(
            'Specificatie',
            'Regel 1: Het wachtwoord bestaat uit minstens 5 karakters.\n',
            'Regel 2: Het wachtwoord bevat een hoofdletter.\n',
            'Regel 3: Het wachtwoord bevat een speciaal teken (#@).\n',
            'Regel 4: De cijfers in het wachtwoord tellen op tot 13.\n',
        )
        self.parameters = [
            Variable(
                'password',
                'str',
                Template('', 'Wachtwoord')
            )
        ]
        self.unit = Variable(
            'is_strong_password',
            'bool',
            Template('', 'Is het een sterk wachtwoord?')
        )
        self.function_elements = [
            [
                'if len(password) < 5: return False',
                'if len(password) <= 5: return False',
                'if len(password) < 8: return False',
                'if len(password) >= 5: return True',
                ''
            ],
            [
                'if not password.isupper(): return False',
                'if password.isupper(): return True',
                ''
            ],
            [
                'if not any(char in "#@" for char in password): return False',
                'if all(char != "#" for char in password): return False',
                'if all(char != "@" for char in password): return False',
                'if not password in "#@": return False',
                'if password and password[0] not in "#@": return False',
                'if password and password[-1] not in "#@": return False',
                'if password and password[0] in "#@": return True',
                'if password and password[-1] in "#@": return True',
                ''
            ],
            [
                'if sum([int(char) for char in password if char.isdigit()]) != 13: return False',
                'if sum([int(char) for char in password if char.isdigit()]) > 13: return False',
                'if sum([int(char) for char in password if char.isdigit()]) < 13: return False',
                'if sum([int(char) for char in password if char.isdigit()]) == 13: return True',
                ''
            ],
            [
                'return True',
            ],
        ]
        self.special_unit_tests = [
            UnitTest(['A346#'], True),
            UnitTest(['@2551B'], True),
            UnitTest(['@34D52'], False),
            UnitTest(['1#236D0'], False),
            UnitTest(['85EFG'], False),
            UnitTest(['@9#4@'], False),
            UnitTest(['@67B'], False),
        ]

    def __generate_digits(self):
        for d1 in range(10):
            for d2 in range(d1, 10):
                if d1 + d2 == 13:
                    yield [d1, d2]
                for d3 in range(d2, 10):
                    if d1 + d2 + d3 == 13:
                        yield [d1, d2, d3]
                    for d4 in range(d3, 10):
                        if d1 + d2 + d3 + d4 == 13:
                            yield [d1, d2, d3, d4]

    def __generate_letters(self):
        for u1 in 'ABCDEFGHIJKLMNOPQRSTUVWXYZ':
            yield [u1]
            for u2 in 'ABCDEFGHIJKLMNOPQRSTUVWXYZ':
                yield []
                yield [u1, u2]

    def __generate_special_characters(self):
        for s1 in '@#!':
            yield [s1]
            for s2 in '@#!':
                yield []
                yield [s1, s2]

    def general_arguments_generator(self):
        digits = list(self.__generate_digits())
        letters = list(self.__generate_letters())
        special_characters = list(self.__generate_special_characters())
        for _ in range(100):
            ds = random.choice(digits)
            us = random.choice(letters)
            scs = random.choice(special_characters)
            chars = [str(d) for d in ds] + us + scs
            random.shuffle(chars)
            yield [''.join(chars)]

            pos = random.randint(0, len(chars) - 1)
            chars[pos] = random.choice('0123456789')
            yield [''.join(chars)]

File: game_dutch_wachtwoord.js
"use strict";
class Wachtwoord extends Dutch {
    constructor() {
        super();
    }
    description() {
        return 'Bepaal of een wachtwoord voldoende sterk is';
    }
    introductionMessage(initialScore, penaltyBug) {
        return new Section([
            new Paragraph('Een webshop laat een nieuwe functie ontwikkelen door een extern softwarebedrijf. ' +
                'De functie wordt gebruikt om te controleren of een wachtwoord sterk genoeg is. ' +
                'Gebruikers moeten erop kunnen rekenen dat hun gegevens goed beschermd zijn. ' +
                'Daarom hebben ze jou ingehuurd om voldoende unit testen te schrijven voor die functie. ' +
                'zodat de functie geen foute resultaten meer kan geven. ' +
                `Voor het hele traject krijg je ${this.formatScore(initialScore)}. ` +
                'In het menu staan bij sommige acties kosten vermeld voor jou. ' +
                'Als een gebruiker bijvoorbeeld een fout constateert in een functie die slaagt voor al jouw unit testen, ' +
                `dan betaal jij een boete van ${this.formatScore(penaltyBug)}.`),
        ]);
    }
    specificationPanel() {
        return new Section([
            new Header('Specificatie'),
            new Paragraph('Regel 1: Het wachtwoord bestaat uit minstens 5 karakters.'),
            new Paragraph('Regel 2: Het wachtwoord bevat een hoofdletter.'),
            new Paragraph('Regel 3: Het wachtwoord bevat een speciaal teken ("#" of "@").'),
            new Paragraph('Regel 4: De cijfers in het wachtwoord tellen op tot 13.'),
        ]);
    }
    getParameters() {
        return [
            new TextVariable('Wachtwoord', 'password')
        ];
    }
    getUnit() {
        return new CheckboxVariable('Is het wachtwoord sterk?', 'isStrong');
    }
    getCandidateElements() {
        return [
            [
                'if (password.length < 5) return false',
                'if (password.length <= 5) return false',
                'if (password.length < 8) return false',
                'if (password.length >= 5) return true',
                '',
            ],
            [
                'if (!/[A-Z]/.test(password)) return false',
                'if (/[A-Z]/.test(password)) return true',
                '',
            ],
            [
                'if (!/[#@]/.test(password)) return false',
                'if (!password.includes("#")) return false',
                'if (!password.includes("@")) return false',
                'if (!["#", "@"].includes(password[0])) return false',
                'if (!["#", "@"].includes(password[password.length - 1])) return false',
                'if (["#", "@"].includes(password[0])) return true',
                'if (["#", "@"].includes(password[password.length - 1])) return true',
                '',
            ],
            [
                'if (password.split("").reduce((sum, char) => sum + (parseInt(char) || 0), 0) !== 13) return false',
                'if (password.split("").reduce((sum, char) => sum + (parseInt(char) || 0), 0) > 13) return false',
                'if (password.split("").reduce((sum, char) => sum + (parseInt(char) || 0), 0) < 13) return false',
                'if (password.split("").reduce((sum, char) => sum + (parseInt(char) || 0), 0) === 13) return true',
                '',
            ],
            [
                'return true',
                'return false',
            ],
        ];
    }
    getSpecialUnitTests() {
        return [
            new UnitTest(['A346#'], true),
            new UnitTest(['@2551B'], true),
            new UnitTest(['@34D52'], false),
            new UnitTest(['1#236D0'], false),
            new UnitTest(['85EFG'], false),
            new UnitTest(['@9#4@'], false),
            new UnitTest(['@67B'], false),
        ];
    }
    *generateDigits() {
        for (let d1 = 0; d1 < 10; d1++) {
            for (let d2 = d1; d2 < 10; d2++) {
                if (d1 + d2 === 13)
                    yield [d1.toString(), d2.toString()];
                for (let d3 = d2; d3 < 10; d3++) {
                    if (d1 + d2 + d3 === 13)
                        yield [d1.toString(), d2.toString(), d3.toString()];
                    for (let d4 = d3; d4 < 10; d4++) {
                        if (d1 + d2 + d3 + d4 === 13)
                            yield [d1.toString(), d2.toString(), d3.toString(), d4.toString()];
                    }
                }
            }
        }
    }
    *generateLetters() {
        const uppercase = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
        for (const u1 of uppercase) {
            yield [u1];
            for (const u2 of uppercase) {
                yield [];
                yield [u1, u2];
            }
        }
    }
    *generateSpecialCharacters() {
        const specialChars = "@#!";
        for (const s1 of specialChars) {
            yield [s1];
            for (const s2 of specialChars) {
                yield [];
                yield [s1, s2];
            }
        }
    }
    *generalArgumentsGenerator() {
        const digits = [...this.generateDigits()];
        const letters = [...this.generateLetters()];
        const specialChars = [...this.generateSpecialCharacters()];
        for (let i = 0; i < 100; i++) {
            const ds = digits.random();
            const us = letters.random();
            const scs = specialChars.random();
            const chars = [...ds, ...us, ...scs];
            this.shuffleArray(chars);
            yield [chars.join("")];
            const pos = Math.floor(Math.random() * chars.length);
            chars[pos] = Math.floor(Math.random() * 10).toString();
            yield [chars.join("")];
        }
    }
    shuffleArray(array) {
        for (let i = array.length - 1; i > 0; i--) {
            const j = Math.floor(Math.random() * (i + 1));
            const swap = array[i];
            array[i] = array[j];
            array[j] = swap;
        }
    }
}
